src/psyclone/core/symbolic_maths.py

```diff
@@ -0,0 +1,110 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author J. Henrichs, Bureau of Meteorology
+# -----------------------------------------------------------------------------
+
+''' This module provides access to sympy-based symbolic maths
+functions.'''
+
+from sympy import simplify
+from sympy.parsing.sympy_parser import parse_expr
+
+
+class SymbolicMaths:
+    '''A wrapper around the symbolic maths package 'sympy'. It
+    provides convenience functions for PSyclone. It has a Singleton
+    access, e.g.:
+
+    >>> from psyclone.psyir.backend.fortran import FortranWriter
+    >>> from psyclone.core import SymbolicMaths
+    >>> sympy = SymbolicMaths.get()
+    >>> # Assume lhs is the PSyIR of 'i+j', and rhs is 'j+i'
+    >>> if sympy.equal(lhs, rhs):
+    ...     writer = FortranWriter()
+    ...     print("'{0}' and '{1}' are equal."
+    ...           .format(writer(lhs), writer(rhs)))
+    'i + j' and 'j + i' are equal.
+
+    '''
+    # Keeps track if importing sympy has been tried.
+    _has_been_imported = False
+
+    # Class variable to store the SymbolicMaths instance if sympy is
+    # available, or None otherwise.
+    _instance = None
+
+    # -------------------------------------------------------------------------
+    @staticmethod
+    def get():
+        '''Static function that if necessary creates and returns the singleton
+        SymbolicMaths instance.
+
+        :returns: the instance of the symbolic maths class.
+        :rtype: :py:class:`psyclone.core.SymbolicMaths.`
+
+        '''
+        if SymbolicMaths._instance is None:
+            SymbolicMaths._instance = SymbolicMaths()
+
+        return SymbolicMaths._instance
+
+    # -------------------------------------------------------------------------
+    def __init__(self):
+
+        # Avoid circular import
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.backend.sympy_writer import SymPyWriter
+
+        self._writer = SymPyWriter()
+
+    # -------------------------------------------------------------------------
+
+    def equal(self, exp1, exp2):
+        '''Test if the two PSyIR operations are identical. This is
+        done by converting the operations to the equivalent Fortran
+        representation, which can be fed into sympy for evaluation.
+
+        :param exp1: the first expression to be compared.
+        :type exp1: py:class:`psyclone.psyir.nodes.Node` or None
+        :param exp2: the first expression to be compared.
+        :type exp2: py:class:`psyclone.psyir.nodes.Node` or None
+
+        '''
+        # Some tests provide a None as parameters
+        if exp1 is None or exp2 is None:
+            return exp1 == exp2
+
+        str_exp1 = parse_expr(self._writer(exp1))
+        str_exp2 = parse_expr(self._writer(exp2))
+        return simplify(str_exp1 == str_exp2)
```

Comment:
I've added a comment to explain this (in short, there are several ways that sympy evaluates expression, and 'simplify' is a mixture/order of operations that works well - it's not clearly specified exactly that it does, and it warns that it might change from one version to the other). In my tests it does exactly what I wanted, without the need to find our own best way of applying operations.

Reply:
OK, thanks for explaining.